libgadget/plane.c

```diff
@@ -0,0 +1,279 @@
+
+        // plane thickness
+        PlaneParams.Thickness = param_get_double(ps, "PlaneThickness");
+        
+    }
+    MPI_Bcast(&PlaneParams, sizeof(struct plane_params), MPI_BYTE, 0, MPI_COMM_WORLD);
+}
+
+void write_plane(int snapnum, const double atime, const Cosmology * CP, const char * OutputDir, const double UnitVelocity_in_cm_per_s, const double UnitLength_in_cm) {
+
+    // int snapnum = get_snap_number(OutputDir);
+        // simulation parameters and variables
+    double BoxSize = PartManager->BoxSize;
+
+    int64_t num_particles_tot = 0; // number of dark matter particles
+    // Use MPI_Allreduce to get the total number of particles on all ranks
+    MPI_Allreduce(&PartManager->NumPart, &num_particles_tot, 1, MPI_INT64, MPI_SUM, MPI_COMM_WORLD);
+    // printf("Total number of particles: %ld\n", num_particles_tot);
+
+    // plane parameters
+    int plane_resolution = PlaneParams.Resolution;
+    double thickness = PlaneParams.Thickness; // in kpc/h
+
+
+    int ThisTask;
+    MPI_Comm_rank(MPI_COMM_WORLD, &ThisTask);
+    
+    double redshift = 1./atime - 1.;
+    message(0, "Computing and writing potential planes.\n");
+
+    double **plane_result = allocate_2d_array(plane_resolution, plane_resolution);
+
+    double **summed_plane_result = NULL;
+        
+    if(ThisTask == 0) {
+        summed_plane_result = allocate_2d_array(plane_resolution, plane_resolution);
+    }
+
+    double comoving_distance = compute_comoving_distance(CP, atime, 1., UnitVelocity_in_cm_per_s);
+
+    // print comoving distance
+    message(0, "Comoving distance: %g\n", comoving_distance);
+
+    /* loop over cut points and normal directions to generate lensing potential planes */
+    for (int i = 0; i < PlaneParams.CutPointsLength; i++) {
+        for (int j = 0; j < PlaneParams.NormalsLength; j++) {
+            MPI_Barrier(MPI_COMM_WORLD);
+
+            message(0, "Computing for cut %d and normal %d\n", i, PlaneParams.Normals[j]);
+
+            // Initialize lensing_potential with zeros
+            for (int i = 0; i < plane_resolution; i++) {
+                for (int j = 0; j < plane_resolution; j++) {
+                    plane_result[i][j] = 0.0;  // Initially zero
+                }
+            }
+
+            double left_corner[3] = {0, 0, 0};
+            int64_t num_particles_plane = 0, num_particles_plane_tot = 0;
+            // print input parameters
+
+            
+
+            /*computing lensing potential planes*/
+            num_particles_plane = cutPlaneGaussianGrid(num_particles_tot,  comoving_distance, BoxSize, CP, atime, PlaneParams.Normals[j], PlaneParams.CutPoints[i], thickness, left_corner, plane_resolution, plane_result);
+            
+            /*sum up planes from all tasks*/
+            for (int k = 0; k < plane_resolution; k++) {
+            MPI_Reduce(plane_result[k], (ThisTask == 0 ? summed_plane_result[k] : NULL), plane_resolution, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
```

Comment:
Should be :
MPI_Reduce(plane_result, summed_plane_result, plane_resolution, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
and no need for the loop, it reduces the whole array.

Reply:
Oh, I see, you are trying to reduce a 2D array! Maybe easier to use a flat array of memory so you can have a single reduce. Multiple reduces in a loop sometimes cause MPI to break.